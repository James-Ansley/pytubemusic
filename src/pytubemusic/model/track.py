"""
Data types for collections of tracks
"""
from typing import Literal

from pydantic import Field

from .base import MaybeTimedelta, Model
from .cover import MaybeCover
from .tags import TrackTags

__all__ = (
    "Track",
    "Single",
    "Split",
    "Playlist",
    "MergePlaylist",
    "Merge",
    "TrackStub",
    "AudioStub",
    "TimeStub",
)

type Track = Single | Split | Playlist | MergePlaylist | Merge


class Single(Model):
    """A single track"""

    __match_args__ = ("url", "metadata", "cover", "start", "end")

    url: str = Field(pattern=r"/watch\?v=")
    metadata: TrackTags
    cover: MaybeCover = None
    start: MaybeTimedelta = None
    end: MaybeTimedelta = None


class TrackStub(Model):
    """A track that exists as a segment of a larger track"""

    __match_args__ = ("metadata", "start", "end")

    metadata: TrackTags
    start: MaybeTimedelta = None
    end: MaybeTimedelta = None


class AudioStub(Model):
    """A track used only for audio with no associated metadata"""

    __match_args__ = ("url", "start", "end")

    url: str = Field(pattern=r"/watch\?v=")
    start: MaybeTimedelta = None
    end: MaybeTimedelta = None


class TimeStub(Model):
    """Timestamp information"""

    __match_args__ = ("url", "start", "end")

    start: MaybeTimedelta = None
    end: MaybeTimedelta = None


class Split(Model):
    """A single track that is split into smaller individual tracks"""

    __match_args__ = ("url", "cover", "tracks")

    url: str = Field(pattern=r"/watch\?v=")
    cover: MaybeCover = None
    tracks: tuple[TrackStub, ...] = Field(min_length=1)


class Playlist(Model):
    """A playlist"""

    __match_args__ = ("url", "cover", "tracks")

    url: str = Field(pattern=r"/playlist\?list=")
    cover: MaybeCover = None
    tracks: tuple[TrackStub | Literal["DROP"], ...] | None = None


class MergePlaylist(Model):
    """Playlist tracks combined into one"""

    __match_args__ = ("url", "metadata", "cover", "tracks")

    url: str = Field(pattern=r"/playlist\?list=")
    metadata: TrackTags
    cover: MaybeCover = None
    tracks: tuple[TimeStub | Literal["DROP"], ...] | None = None


class Merge(Model):
    """Multiple tacks combined into one"""

    __match_args__ = ("metadata", "cover", "tracks")

    metadata: TrackTags
    cover: MaybeCover = None
    tracks: tuple[AudioStub, ...] = Field(min_length=1)
