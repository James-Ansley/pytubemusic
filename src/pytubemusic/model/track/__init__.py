from .track_data import *
