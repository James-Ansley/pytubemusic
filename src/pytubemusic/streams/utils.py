from contextlib import AbstractContextManager, contextmanager
from typing import IO


@contextmanager
def stream[T: IO](buff: T) -> AbstractContextManager[T]:
    try:
        yield buff
    finally:
        buff.seek(0)
