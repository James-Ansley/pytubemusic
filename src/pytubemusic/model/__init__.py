import pydantic as _pydantic

from .base import *
from .album import *
from .cover import *
from .tags import *
from .track import *
from .track_data import *

type RootType = "Album | Track"
RootModel = _pydantic.RootModel[RootType]
