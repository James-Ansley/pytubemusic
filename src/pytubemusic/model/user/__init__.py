"""
Data types for user-entered data
"""
from .album import *
from .cover import *
from .root import *
from .tags import *
from .track import *
