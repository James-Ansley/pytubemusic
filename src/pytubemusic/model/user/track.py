"""
Data types for collections of tracks
"""
from typing import Literal

from pydantic import Field, RootModel

from pytubemusic.model.types import MaybeTimedelta
from .base import Model
from .cover import MaybeCover
from .tags import TrackTags

__all__ = (
    "Track",
    "TrackType",
    "Single",
    "Split",
    "Playlist",
    "Merge",
    "TrackStub",
    "AudioStub",
    "TimeStub",
    "MergeStub",
    "Drop",
)

type TrackType = Single | Split | Playlist | Merge


class Track:
    def __new__(cls, **kwargs) -> TrackType:
        # noinspection PyTypeChecker
        return RootModel[TrackType](**kwargs).root


class Single(Model):
    """A single track"""

    __match_args__ = ("url", "metadata", "cover", "start", "end")

    url: str = Field(pattern=r"/watch\?v=")
    metadata: TrackTags
    cover: MaybeCover = None
    start: MaybeTimedelta = None
    end: MaybeTimedelta = None


class TrackStub(Model):
    """A track that exists as a segment of a larger track"""

    __match_args__ = ("metadata", "cover", "start", "end")

    metadata: TrackTags
    cover: MaybeCover = None
    start: MaybeTimedelta = None
    end: MaybeTimedelta = None


class Drop(Model):
    __match_args__ = ("drop",)

    drop: Literal[True] = True


class AudioStub(Model):
    """A track used only for audio with no associated metadata"""

    __match_args__ = ("url", "start", "end")

    url: str = Field(pattern=r"/watch\?v=")
    start: MaybeTimedelta = None
    end: MaybeTimedelta = None


class TimeStub(Model):
    """Timestamp information"""

    __match_args__ = ("url", "start", "end")

    start: MaybeTimedelta = None
    end: MaybeTimedelta = None


class MergeStub(Model):
    """Multiple time stamps clamped into one"""

    __match_args__ = ("metadata", "cover", "parts")

    metadata: TrackTags
    cover: MaybeCover = None
    parts: tuple[TimeStub | Drop, ...]


class Split(Model):
    """A single track that is split into smaller individual tracks"""

    __match_args__ = ("url", "cover", "tracks")

    url: str = Field(pattern=r"/watch\?v=")
    cover: MaybeCover = None
    tracks: tuple[TrackStub, ...] = Field(min_length=1)


class Playlist(Model):
    """A playlist"""

    __match_args__ = ("url", "cover", "tracks")

    url: str = Field(pattern=r"/playlist\?list=")
    cover: MaybeCover = None
    tracks: tuple[TrackStub | MergeStub | Drop, ...] = ()


class Merge(Model):
    """Multiple tacks combined into one"""

    __match_args__ = ("metadata", "cover", "tracks")

    metadata: TrackTags
    cover: MaybeCover = None
    parts: tuple[AudioStub, ...] = Field(min_length=1)
