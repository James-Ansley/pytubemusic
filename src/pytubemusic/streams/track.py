from functools import reduce
from pathlib import Path

from pytubemusic.model.audio import Audio, RawAudio
from pytubemusic.model.track import TrackData
from pytubemusic.streams.audio import fetch_audio_data
from pytubemusic.streams.images import fetch_cover_data


def fetch_track(track: TrackData, context: Path = None) -> Audio:
    return Audio(
        raw_audio=reduce(
            merge_audio,
            (fetch_audio_data(part) for part in track.parts),
        ),
        metadata=track.metadata,
        cover=fetch_cover_data(track.cover, context),
    )


def merge_audio(audio1: RawAudio, audio2: RawAudio) -> RawAudio:
    return RawAudio(
        segment=audio1.segment + audio2.segment,
        bit_rate=max(audio1.bit_rate, audio2.bit_rate),
    )
