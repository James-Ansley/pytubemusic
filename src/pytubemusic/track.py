import os
from collections.abc import Iterable, Mapping
from dataclasses import dataclass
from datetime import timedelta
from pathlib import Path
from tempfile import NamedTemporaryFile
from typing import Any, Self

from pipe_utils import Pipe
from pytube import Playlist

from pytubemusic.audio import Audio
from pytubemusic.logutils import log_call
from pytubemusic.utils import *

__all__ = ["Track"]

STR_MAP = Mapping[str, Any]
TRACK_DATA = Iterable[STR_MAP]


@dataclass
class Track:
    raw_audio: Audio
    cover: NamedTemporaryFile
    start: timedelta
    end: timedelta
    metadata: STR_MAP

    @property
    def audio(self) -> Audio:
        """The raw audio snipped to the track start and end times"""
        return self.raw_audio.snip(self.start, self.end)

    @property
    def title(self) -> str:
        """The track title"""
        return self.metadata["title"]

    @property
    def album(self) -> str | None:
        """The track album â€“ May be None"""
        return self.metadata.get("album")

    @log_call(on_enter="Exporting '{self.metadata[title]}' to '{folder}'")
    def export(self, folder: Path = Path(".")):
        """
        Writes the track to folder / {track.title}.mp3

        Creates any dirs necessary in the folder path.
        Overwrites existing tracks with the same name.

        :param folder: The directory the track will be written to
        """
        os.makedirs(folder, exist_ok=True)
        with open(pathify(folder, self.title), "wb") as f:
            self.audio.export(f, self.metadata, self.cover.name)

    @classmethod
    @log_call(on_enter="Downloading '{metadata[album]}' from {url}")
    def from_album(
            cls,
            url: str,
            *,
            track_data: Iterable[Mapping[str, Any]],
            metadata: Mapping[str, str],
            cover_url: str = None,
    ) -> Iterable[Self]:
        """
        Factory that constructs an Album from a single video that contains
        multiple tracks in sequence.

        ``track_data`` is a list of string maps indicating track-specific
        data in the order of the tracks in the video and is of the form::

            [ { start: ..., [end: ...], metadata: { title: ..., ... } }, ... ]

        Where ``start`` and ``end`` (optional) are ``H?:M:S`` timestamps that
        indicate the start and end time stamps of the track and ``metadata``
        is a string map of track-specific FFMPEG MP3 metadata tags.

        :param url: The URL of the video
        :param track_data: A list of string maps of track data
        :param metadata: A String map of FFMPEG MP3 metadata tags
        :param cover_url: The url to a JPG cover image
        :return: an iterator of Track objects
        """
        audio = Audio.from_url(url)
        track_data = (
                Pipe(track_data)
                | (set_ends, audio.duration)
                | (merge_metadata, metadata)
        ).get()
        for track in track_data:
            yield cls(
                audio,
                thumbnail(url) if cover_url is None else get_cover(cover_url),
                to_delta(track["start"]),
                to_delta(track["end"]),
                track["metadata"],
            )

    @classmethod
    @log_call(on_enter="Downloading '{metadata[title]}' from {url}")
    def from_video(
            cls,
            url: str,
            *,
            metadata: Mapping[str, str],
            start: str = "00:00",
            end: str = None,
            cover_url: str = None,
    ) -> Self:
        """
        Factory that constructs a Track from a video.

        :param url: The URL of the video
        :param metadata: A String map of FFMPEG MP3 metadata tags
        :param start: A H?:M:S.f? timestamp indicating the start time of the
            track. Defaults to 00:00:00.00
        :param end: A H?:M:S.f? timestamp indicating the end time of the track.
            Defaults to the length of the audio
        :param cover_url: The url to a JPG cover image
        :return: A new Track
        """
        audio = Audio.from_url(url)
        return cls(
            audio,
            thumbnail(url) if cover_url is None else get_cover(cover_url),
            to_delta(start),
            to_delta(end) if end is not None else audio.duration,
            metadata,
        )

    @classmethod
    @log_call(on_enter="Downloading '{metadata[album]}' from playlist {url}")
    def from_playlist(
            cls,
            url: str,
            *,
            metadata: Mapping[str, str],
            track_data: Iterable[Mapping[str, Any]] = None,
            cover_url: str = None,
    ) -> Iterable[Self]:
        """
        Factory that constructs an Album from a playlist video that contains
        multiple tracks in sequence.

        :param url: The URL of the playlist
        :param track_data: A list of string maps of track data
        :param metadata: A String map of FFMPEG MP3 metadata tags
        :param cover_url: The url to a JPG cover image
        :return: an iterator of Track objects
        """
        playlist = Playlist(url)
        track_data = [{}] * len(playlist) if track_data is None else track_data
        video_data = zip(playlist.videos, track_data)
        for i, (video, data) in enumerate(video_data, start=1):
            yield cls.from_video(
                video.watch_url,
                cover_url=cover_url,
                metadata={
                    **metadata,
                    "track": i,
                    "title": video.title,
                    **data.get("metadata", {}),
                },
                **{k: v for k, v in data.items() if k != "metadata"},
            )
