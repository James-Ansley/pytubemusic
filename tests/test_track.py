from datetime import timedelta
from pprint import pformat

from approvaltests import verify
from pydantic import ValidationError
from pytest import raises

from pytubemusic.model.user import File, TrackTags, Url
from pytubemusic.model.user.track import (Merge, Playlist,
                                          Single, Split, Track)
from tests import test
from tests.utils import load_toml


@test()
def tracks_cannot_be_created_from_empty_data():
    raises(ValidationError, lambda: Track(**{}))


@test(
    depends_on=("test_tags.py::track_tags_can_be_created_with_minimal_data",),
    scope="session",
)
def singles_can_be_created_with_minimal_data():
    data = load_toml("single_minimal.toml")
    track = Track(**data)
    assert isinstance(track, Single)
    verify(pformat(track))


@test(depends_on=("singles_can_be_created_with_minimal_data",))
def single_urls_must_have_a_watch_id():
    data = {"url": "www.example.com", "metadata": {"title": "My Track Title"}}
    raises(ValidationError, lambda: Single(**data))


@test(depends_on=("singles_can_be_created_with_minimal_data",))
def single_metadata_must_include_track_tag_data():
    data = {
        "url": "www.example.com/watch?v=",
        "metadata": {"album": "My Album Title"}
    }
    raises(ValidationError, lambda: Single(**data))


@test(depends_on=("singles_can_be_created_with_minimal_data",))
def singles_must_not_contain_extra_fields():
    data = {
        "url": "www.example.com/watch?v=",
        "metadata": {"title": "My Track Title"},
        "foo": "bar",
    }
    raises(ValidationError, lambda: Single(**data))


@test(depends_on=("singles_can_be_created_with_minimal_data",))
def singles_must_have_a_url_and_metadata():
    raises(ValidationError, lambda: Single(**{
        "url": "www.example.com/watch?v=",
    }))
    raises(ValidationError, lambda: Single(**{
        "metadata": {"title": "My Track Title"},
    }))


@test(
    depends_on=(
          "test_track.py::singles_can_be_created_with_minimal_data",
          "test_cover.py::url_covers_can_be_created_with_a_hyperlink",
          "test_cover.py::file_covers_can_be_created_with_a_path",
    ),
    scope="session",
)
def singles_can_have_file_or_url_covers():
    data = {
        "url": "www.example.com/watch?v=",
        "metadata": {"title": "My Track Title"}
    }
    track = Single(**data | {"cover": {"path": "data/pic.jpeg"}})
    assert track.cover == File(**{"path": "data/pic.jpeg"})
    track = Single(**data | {"cover": {"href": "www.example.com/pic.jpeg"}})
    assert track.cover == Url(**{"href": "www.example.com/pic.jpeg"})


@test(depends_on=("singles_can_be_created_with_minimal_data",))
def singles_can_be_created_with_start_and_end_times():
    data = {
        "url": "www.example.com/watch?v=",
        "metadata": {"title": "My Track Title"}
    }
    track1 = Single(**data | {"start": "00:00:01.5"})
    track2 = Single(**data | {"end": "00:03:21"})
    track3 = Single(**data | {"start": "00:00:01.5", "end": "00:03:21.0"})
    assert track1.start == timedelta(seconds=1, milliseconds=500)
    assert track2.end == timedelta(minutes=3, seconds=21)
    assert track3.start == timedelta(seconds=1, milliseconds=500)
    assert track3.end == timedelta(minutes=3, seconds=21)


@test(
    depends_on=(
          "test_track.py::singles_can_be_created_with_minimal_data",
          "test_cover.py::url_covers_can_be_created_with_a_hyperlink",
          "test_cover.py::file_covers_can_be_created_with_a_path",
    ),
    scope="session",
)
def singles_can_be_created_with_full_data():
    data = load_toml("single_full.toml")
    single = Single(**data)
    verify(pformat(single))


@test(
    depends_on=("test_tags.py::track_tags_can_be_created_with_minimal_data",),
    scope="session",
)
def split_tracks_can_be_created_with_minimal_data():
    data = load_toml("split_track_minimal.toml")
    split = Track(**data)
    assert isinstance(split, Split)
    verify(pformat(split))


@test(depends_on=("split_tracks_can_be_created_with_minimal_data",))
def split_tracks_cannot_be_created_with_empty_track_lists():
    raises(
        ValidationError,
        lambda: Split(**{"url": "www.example.com/watch?v=", "tracks": []}),
    )


@test(depends_on=("split_tracks_can_be_created_with_minimal_data",))
def split_track_urls_must_have_a_watch_id():
    data = {
        "url": "www.example.com",
        "tracks": [{"metadata": {"title": "My Track Title"}}],
    }
    raises(ValidationError, lambda: Split(**data))


@test(depends_on=("split_tracks_can_be_created_with_minimal_data",))
def split_tracks_must_not_contain_extra_fields():
    data = {
        "url": "www.example.com/watch?v=",
        "tracks": [{"metadata": {"title": "My Track Title"}}],
        "foo": "bar",
    }
    raises(ValidationError, lambda: Split(**data))


@test(
    depends_on=(
          "test_track.py::split_tracks_can_be_created_with_minimal_data",
          "test_cover.py::url_covers_can_be_created_with_a_hyperlink",
          "test_cover.py::file_covers_can_be_created_with_a_path",
    ),
    scope="session",
)
def split_tracks_can_be_created_with_full_data():
    data = load_toml("split_track_full.toml")
    single = Split(**data)
    verify(pformat(single))


@test()
def playlists_can_be_created_with_minimal_data():
    data = load_toml("playlist_minimal.toml")
    playlist = Track(**data)
    assert isinstance(playlist, Playlist)
    assert playlist.url == data["url"]
    assert playlist.cover is None
    assert playlist.tracks == ()


@test(depends_on=("playlists_can_be_created_with_minimal_data",))
def playlists_must_have_a_valid_id():
    data = {"url": "www.example.com"}
    raises(ValidationError, lambda: Playlist(**data))


@test(depends_on=("playlists_can_be_created_with_minimal_data",))
def playlists_must_not_contain_extra_fields():
    data = {
        "url": "www.example.com/playlist?list=",
        "foo": "bar",
    }
    raises(ValidationError, lambda: Playlist(**data))


@test(
    depends_on=(
          "test_track.py::playlists_can_be_created_with_minimal_data",
          "test_tags.py::track_tags_can_be_created_with_minimal_data",
    ),
    scope="session",
)
def playlist_tracks_can_be_dropped():
    data = {
        "url": "www.example.com/playlist?list=",
        "tracks": [
            {"metadata": {"title": "My Track Title"}},
            {"drop": True},
            {"metadata": {"title": "My Other Track Title"}},
        ]
    }
    playlist = Playlist(**data)
    verify(pformat(playlist))


@test(
    depends_on=(
          "test_track.py::playlists_can_be_created_with_minimal_data",
          "test_tags.py::track_tags_can_be_created_with_minimal_data",
          "test_cover.py::url_covers_can_be_created_with_a_hyperlink",
          "test_cover.py::file_covers_can_be_created_with_a_path",
    ),
    scope="session",
)
def playlist_tracks_can_be_merged():
    data = load_toml("playlist_with_merge.toml")
    playlist = Playlist(**data)
    verify(pformat(playlist))


@test(
    depends_on=(
          "test_track.py::playlists_can_be_created_with_minimal_data",
          "test_tags.py::track_tags_can_be_created_with_minimal_data",
          "test_cover.py::url_covers_can_be_created_with_a_hyperlink",
          "test_cover.py::file_covers_can_be_created_with_a_path",
    ),
    scope="session",
)
def playlists_can_be_created_with_full_data():
    data = load_toml("playlist_full.toml")
    playlist = Track(**data)
    verify(pformat(playlist))


@test(
    depends_on=("test_tags.py::track_tags_can_be_created_with_minimal_data",),
    scope="session",
)
def merge_tracks_can_be_created_with_minimal_data():
    data = load_toml("merge_minimal.toml")
    merge = Track(**data)
    assert isinstance(merge, Merge)
    assert merge.metadata == TrackTags(**data["metadata"])
    assert len(merge.parts) == 1
    assert merge.parts[0].url == "www.example.com/watch?v="


@test(depends_on=("merge_tracks_can_be_created_with_minimal_data",))
def merge_tracks_must_contain_at_least_one_track():
    data = {"metadata": {"title": "My Track Title"}, "parts": []}
    raises(ValidationError, lambda: Merge(**data))


@test(depends_on=("merge_tracks_can_be_created_with_minimal_data",))
def merge_track_urls_must_contain_a_valid_id():
    data = {
        "metadata": {"title": "My Track Title"},
        "parts": [{"url": "www.example.com"}]
    }
    raises(ValidationError, lambda: Merge(**data))


@test(depends_on=("merge_tracks_can_be_created_with_minimal_data",))
def merge_tracks_cannot_contain_extra_fields():
    data = {
        "metadata": {"title": "My Track Title"},
        "parts": [{"url": "www.example.com/watch?v="}],
        "foo": "bar"
    }
    raises(ValidationError, lambda: Merge(**data))


@test(
    depends_on=(
          "test_track.py::merge_tracks_can_be_created_with_minimal_data",
          "test_cover.py::url_covers_can_be_created_with_a_hyperlink",
          "test_cover.py::file_covers_can_be_created_with_a_path",
    ),
    scope="session",
)
def merge_tracks_can_be_created_with_full_data():
    data = load_toml("merge_full.toml")
    merge = Track(**data)
    verify(pformat(merge))
