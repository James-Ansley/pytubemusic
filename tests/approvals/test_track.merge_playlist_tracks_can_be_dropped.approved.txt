MergePlaylist(url='www.example.com/playlist?list='
              metadata=TrackTags(title='My Track Title')
              tracks=(TimeStub(start=datetime.timedelta(seconds=1, microseconds=500000)
                               end=datetime.timedelta(seconds=2601)),
                      Drop(drop=True),
                      TimeStub(start=datetime.timedelta(seconds=12, microseconds=300000)
                               end=datetime.timedelta(seconds=601))))
